src/journal/journald-context.c

```diff
@@ -268,6 +274,54 @@ static int client_context_read_label(
         return 0;
 }
 
+static int client_context_read_log_filter_xattr(const char *cgroup, Set **patterns, const char *key) {
+        _cleanup_free_ char *xattr = NULL;
+        int r;
+
+        r = cg_get_xattr_malloc(SYSTEMD_CGROUP_CONTROLLER, cgroup, key, &xattr);
+        if (r == -ENODATA) {
+                /* -ENODATA is returned when the key is not defined, this is not an error. */
+                *patterns = set_free_with_destructor(*patterns, pattern_free);
+                return 0;
+        }
+        if (r < 0)
+                return log_debug_errno(r, "Failed to get %s xattr for %s: %m", key, cgroup);
+
+        const char *pattern;
+        NULSTR_FOREACH(pattern, xattr) {
+                pcre2_code *regex = NULL;
+
+                r = pattern_compile_and_log(pattern, 0, &regex);
+                if (r < 0)
+                        return log_debug_errno(r, "Failed to compile log filtering pattern '%s' for %s: %m",
+                                               pattern, cgroup);
```

Comment:
so with this you fail the whole function leaving the set half-initialized. that is less than ideal.

we should try to make sure functions either fully succeed or fully fail, but this one can leave half-initialized stuff around, which we try to avoid (sometimes it's very hard avoid, so this is not a strict rule, but in this occasion here I think it is easy to avoid).

Quite frankly, i'd just always alocate a new set object here and return that. That we we can have this "atomic" behaviour. We either give you a full new, consistent set on success, or none on failure, but we never modify the existing state half-way when we fail